File: adt2unit/calc_following_order.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

int16_t calc_following_order (uint8_t order)
{
  int16_t r = -1; // result
  uint8_t i = order; // index
  uint8_t n = 0; // jumps count

  DBG_ENTER ("calc_following_order");

  do
  {
    uint8_t j = songdata.pattern_order[i];

    if (j < 0x80)
      r = i;
    else
    {
      i = j - 0x80;
      n++;
    }
  } while ((n < 0x80) && (r == -1));

  DBG_LEAVE (); //EXIT //calc_following_order
  return r;
}

File: adt2unit/pas/calc_following_order.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

function calc_following_order(order: Byte): Integer;

var
  result: Integer;
  index,jump_count: Byte;

begin
  _dbg_enter ({$I %FILE%}, 'calc_following_order');

  result := -1;
  index := order;
  jump_count := 0;

  Repeat
    If (songdata.pattern_order[index] < $80) then result := index
    else begin
           index := songdata.pattern_order[index]-$80;
           Inc(jump_count);
         end;
  until (jump_count > $7f) or
        (result <> -1);

  calc_following_order := result;

  _dbg_leave; //EXIT //calc_following_order
end;
